### DISCUSSION TITLE
Help: Using both Await and AddListener methods on a future

### DISCUSSION THREAD

**Discussion by chamrox89** [netty/netty#12092] [Q&A]
Hi,

What happens if both await and AddListener methods use for a single future. I'm working with a code written like that.
Below are the basic flow;
```
Future<Channel> channelFuture = channelPool.acquire();
channelFuture.await(1000, TimeUnit.MILLISECONDS);
if (channelFuture.isSuccess()) {
               // log line 1
                channelFuture.addListener(future -> {
                    if (future.isSuccess()) {
                        // log line 2
                        Channel channel = (Channel) future.getNow();
                        channel.writeAndFlush(message);
                    }
                });
} 
```

I've noticed that randomly the listener method does not get called in this method. Is this because of the use of the await method?
I'm planning to remove the await call here.
I would like to know if await method calls before the AddListener method is there any possibility that the listener does not get called. As I've came across such situations randomly. I want to to know the exact reason why the listener does not get called in this code randomly.
Thanks in advance.

P.S.:
I had a log line (Log line 1) inside of the if block (if (channelFuture.isSuccess()) {) And that was printed in the log file. But not in the log line (Log line 2) which was in the if (future.isSuccess()) { block in the Listener. Simply log line 1 was printed in the log file, but log line 2 was not printed. So the future was a success but the listener was not getting called even the isSuccess returns true.
If it happened because of the redundant Await method, I can happily remove that. But I want to know that for surely if there is any other reason for this.

**Comment by chrisvest** [comment]
After `await(long, TimeUnit)` returns, the future can be in one of three states: completed successfully, failed, or still not completed.
Future listeners are only called for completed futures, so if your await call times out, or the future fails or is cancelled, then your `isSuccess()` call will return `false` and you won't be adding your listener.
Also, while there is no harm to adding a listener to a future that you know has already completed successfully, there is also little point in doing so. Since you know the future completed successfully, you can just directly call `future.getNow().writeAndFlush(message)`, and then attach the listener in the else-clause.
Or you can omit the outer if-condition entirely and always attach your listener.

**Reply by chamrox89** [reply]
Thanks for your answer. In the case I mentioned about not getting the listener called, the future was in the completed state. Because I had a log line inside of the if block (if (channelFuture.isSuccess()) {) And that was printed in the log file. But not in the log line which was in the if (future.isSuccess()) { block in the Listener.
So the listener was not getting called even the isSuccess returns true..
Because of this incident I would like to know the possible reasons for this. If it happend because of the Await method, I can happily remove that. But I want to know that for surely if there is any other reason for this.

**Comment by chamrox89** [comment]
Update:
I removed the await method and kept only the listener, but when placing a request load (load testing) sometimes it does not getting to the future.isSuccess block. (no log line 2). This is happening very frequently now. It seems it takes forever for the channel acquire method to return a channel and the operation completed method is not getting called. 
Any help is appreciated.
